### ISSUE TITLE
[mlir] Convert-linalg-to-std Pass crashed with "(op->getNumResults() == 0 && "Library call for linalg operation can be generated only for ops that " "have void return types")"

### ISSUE THREAD

**Issue by Colloportus0** [llvm/llvm-project#59987]
MLIR built at commit https://github.com/llvm/llvm-project/commit/a0138390
Reproduced with:  
`mlir-opt --convert-linalg-to-std temp.mlir`

`temp.mlir`:
```mlir
#map = affine_map<(d0, d1) -> (d0, d1)>
#map1 = affine_map<(d0, d1) -> (d0)>
module {
  func.func @func(%arg0: tensor<?x?xf32>, %arg1: tensor<?xf32>)  {
    %0 = linalg.generic {indexing_maps = [#map, #map1], iterator_types = ["parallel", "reduction"]} ins(%arg0 : tensor<?x?xf32>) outs(%arg1 : tensor<?xf32>) {
    ^bb0(%in: f32, %out: f32): 
      linalg.yield %in : f32
    } -> tensor<?xf32>
    return 
  }
}


```
 
trace:
```console  
Assertion failed: (op->getNumResults() == 0 && "Library call for linalg operation can be generated only for ops that " "have void return types"), function getLibraryCallSymbolRef, file LinalgToStandard.cpp, line 74.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: mlir-opt --convert-linalg-to-std temp.mlir
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  mlir-opt                 0x0000000100ba45bc llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 56
1  mlir-opt                 0x0000000100ba3624 llvm::sys::RunSignalHandlers() + 112
2  mlir-opt                 0x0000000100ba4c54 SignalHandler(int) + 344
3  libsystem_platform.dylib 0x00000001a56894c4 _sigtramp + 56
4  libsystem_pthread.dylib  0x00000001a5671ee0 pthread_kill + 288
5  libsystem_c.dylib        0x00000001a55ac340 abort + 168
6  libsystem_c.dylib        0x00000001a55ab754 err + 0
7  mlir-opt                 0x0000000102435d40 mlir::linalg::LinalgOpToLibraryCallRewrite::matchAndRewrite(mlir::linalg::LinalgOp, mlir::PatternRewriter&) const (.cold.15) + 0
8  mlir-opt                 0x00000001019f8fe0 mlir::linalg::LinalgOpToLibraryCallRewrite::matchAndRewrite(mlir::linalg::LinalgOp, mlir::PatternRewriter&) const + 1748
9  mlir-opt                 0x0000000102043bd0 mlir::PatternApplicator::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&, llvm::function_ref<bool (mlir::Pattern const&)>, llvm::function_ref<void (mlir::Pattern const&)>, llvm::function_ref<mlir::LogicalResult (mlir::Pattern const&)>) + 1440
10 mlir-opt                 0x0000000101df14b0 (anonymous namespace)::OperationLegalizer::legalize(mlir::Operation*, mlir::ConversionPatternRewriter&) + 1948
11 mlir-opt                 0x0000000101deab1c (anonymous namespace)::OperationConverter::convertOperations(llvm::ArrayRef<mlir::Operation*>, llvm::function_ref<void (mlir::Diagnostic&)>) + 928
12 mlir-opt                 0x0000000101decdb8 mlir::applyFullConversion(mlir::Operation*, mlir::ConversionTarget&, mlir::FrozenRewritePatternSet const&) + 80
13 mlir-opt                 0x00000001019fa110 (anonymous namespace)::ConvertLinalgToStandardPass::runOnOperation() + 556
14 mlir-opt                 0x0000000101daf4dc mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) + 420
15 mlir-opt                 0x0000000101dafa0c mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) + 320
16 mlir-opt                 0x0000000101db1388 mlir::PassManager::run(mlir::Operation*) + 1148
17 mlir-opt                 0x0000000101daa840 performActions(llvm::raw_ostream&, bool, bool, std::__1::shared_ptr<llvm::SourceMgr> const&, mlir::MLIRContext*, llvm::function_ref<mlir::LogicalResult (mlir::PassManager&)>, bool, bool) + 504
18 mlir-opt                 0x0000000101daa410 mlir::LogicalResult llvm::function_ref<mlir::LogicalResult (std::__1::unique_ptr<llvm::MemoryBuffer, std::__1::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>::callback_fn<mlir::MlirOptMain(llvm::raw_ostream&, std::__1::unique_ptr<llvm::MemoryBuffer, std::__1::default_delete<llvm::MemoryBuffer>>, llvm::function_ref<mlir::LogicalResult (mlir::PassManager&)>, mlir::DialectRegistry&, bool, bool, bool, bool, bool, bool, bool)::$_0>(long, std::__1::unique_ptr<llvm::MemoryBuffer, std::__1::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&) + 704
19 mlir-opt                 0x0000000101e1502c mlir::splitAndProcessBuffer(std::__1::unique_ptr<llvm::MemoryBuffer, std::__1::default_delete<llvm::MemoryBuffer>>, llvm::function_ref<mlir::LogicalResult (std::__1::unique_ptr<llvm::MemoryBuffer, std::__1::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>, llvm::raw_ostream&, bool, bool) + 656
20 mlir-opt                 0x0000000101da8838 mlir::MlirOptMain(llvm::raw_ostream&, std::__1::unique_ptr<llvm::MemoryBuffer, std::__1::default_delete<llvm::MemoryBuffer>>, llvm::function_ref<mlir::LogicalResult (mlir::PassManager&)>, mlir::DialectRegistry&, bool, bool, bool, bool, bool, bool, bool) + 216
21 mlir-opt                 0x0000000101da8d2c mlir::MlirOptMain(int, char**, llvm::StringRef, mlir::DialectRegistry&, bool) + 1208
22 mlir-opt                 0x0000000100a470a0 main + 108
23 dyld                     0x00000001053c9088 start + 516
```

**Comment by nicolasvasilache** [comment]
Filleless repro:
```
cat <<EOF | mlir-opt --convert-linalg-to-std
#map = affine_map<(d0, d1) -> (d0, d1)>
#map1 = affine_map<(d0, d1) -> (d0)>
module {
  func.func @func(%arg0: tensor<?x?xf32>, %arg1: tensor<?xf32>)  {
    %0 = linalg.generic {indexing_maps = [#map, #map1], iterator_types = ["parallel", "reduction"]} ins(%arg0 : tensor<?x?xf32>) outs(%arg1 : tensor<?xf32>) {
    ^bb0(%in: f32, %out: f32): 
      linalg.yield %in : f32
    } -> tensor<?xf32>
    return 
  }
}
EOF
```

**Comment by nicolasvasilache** [comment]
Fixed in c3f0efe753e27105b519ae9283796d41fe574741.

**Comment by nicolasvasilache** [comment]
Missing test added in: 02fb5aae11098fc69ed838eec97f57691eaccfa9
